lightning/src/offers/static_invoice.rs

```diff
@@ -0,0 +1,591 @@
+		WithoutLength(offer_bytes).write(&mut bytes).unwrap();
+		contents.as_invoice_fields_tlv_stream().write(&mut bytes).unwrap();
+
+		let tagged_hash = TaggedHash::from_valid_tlv_stream_bytes(SIGNATURE_TAG, &bytes);
+		Self { contents, tagged_hash, bytes }
+	}
+
+	/// Signs the [`TaggedHash`] of the invoice using the given function.
+	///
+	/// Note: The hash computation may have included unknown, odd TLV records.
+	pub fn sign<F: SignStaticInvoiceFn>(mut self, sign: F) -> Result<StaticInvoice, SignError> {
+		let pubkey = self.contents.signing_pubkey;
+		let signature = merkle::sign_message(sign, &self, pubkey)?;
+
+		// Append the signature TLV record to the bytes.
+		let signature_tlv_stream = SignatureTlvStreamRef { signature: Some(&signature) };
+		signature_tlv_stream.write(&mut self.bytes).unwrap();
+
+		Ok(StaticInvoice {
+			bytes: self.bytes,
+			contents: self.contents,
+			signature,
+			tagged_hash: self.tagged_hash,
+		})
+	}
+
+	invoice_accessors_common!(self, self.contents, StaticInvoice);
+	invoice_accessors!(self, self.contents);
+}
+
+impl AsRef<TaggedHash> for UnsignedStaticInvoice {
+	fn as_ref(&self) -> &TaggedHash {
+		&self.tagged_hash
+	}
+}
+
+/// A function for signing an [`UnsignedStaticInvoice`].
+pub trait SignStaticInvoiceFn {
+	/// Signs a [`TaggedHash`] computed over the merkle root of `message`'s TLV stream.
+	fn sign_invoice(&self, message: &UnsignedStaticInvoice) -> Result<Signature, ()>;
+}
+
+impl<F> SignStaticInvoiceFn for F
+where
+	F: Fn(&UnsignedStaticInvoice) -> Result<Signature, ()>,
+{
+	fn sign_invoice(&self, message: &UnsignedStaticInvoice) -> Result<Signature, ()> {
+		self(message)
+	}
+}
+
+impl<F> SignFn<UnsignedStaticInvoice> for F
+where
+	F: SignStaticInvoiceFn,
+{
+	fn sign(&self, message: &UnsignedStaticInvoice) -> Result<Signature, ()> {
+		self.sign_invoice(message)
+	}
+}
+
+/// A `StaticInvoice` is a reusable payment request corresponding to an [`Offer`].
+///
+/// A static invoice may be sent in response to an [`InvoiceRequest`] and includes all the
+/// information needed to pay the recipient. However, unlike [`Bolt12Invoice`]s, static invoices do
+/// not provide proof-of-payment. Therefore, [`Bolt12Invoice`]s should be preferred when the
+/// recipient is online to provide one.
+///
+/// [`Offer`]: crate::offers::offer::Offer
+/// [`InvoiceRequest`]: crate::offers::invoice_request::InvoiceRequest
+/// [`Bolt12Invoice`]: crate::offers::invoice::Bolt12Invoice
+pub struct StaticInvoice {
+	bytes: Vec<u8>,
+	contents: InvoiceContents,
+	signature: Signature,
+	tagged_hash: TaggedHash,
+}
+
+/// The contents of a [`StaticInvoice`] for responding to an [`Offer`].
+///
+/// [`Offer`]: crate::offers::offer::Offer
+struct InvoiceContents {
+	offer: OfferContents,
+	payment_paths: Vec<(BlindedPayInfo, BlindedPath)>,
+	created_at: Duration,
+	relative_expiry: Option<Duration>,
+	fallbacks: Option<Vec<FallbackAddress>>,
+	features: Bolt12InvoiceFeatures,
+	signing_pubkey: PublicKey,
```

Comment:
These contents are missing the invoice message paths to send the `held_htlc_available` message over, will fix in the next push.